kore/src/Kore/Attribute/Pattern/NonSimplifiable.hs

```diff
@@ -0,0 +1,201 @@
+            NonSimplifiable . Just $ ConstructorLikeHead
+        | otherwise =
+            NonSimplifiable Nothing
+      where
+        child = domainValueChild domainValue
+
+instance Synthetic NonSimplifiable (Equals sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Exists sort variable) where
+    synthetic = const (NonSimplifiable Nothing)
+
+instance Synthetic NonSimplifiable (Floor sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Forall sort variable) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Iff sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Implies sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (In sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Mu sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Next sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Not sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Nu sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Or sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Rewrites sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Builtin key) where
+    synthetic =
+        \case
+            BuiltinInt _    -> NonSimplifiable . Just $ BuiltinHead
+            BuiltinBool _   -> NonSimplifiable . Just $ BuiltinHead
+            BuiltinString _ -> NonSimplifiable . Just $ BuiltinHead
+            _               -> NonSimplifiable Nothing
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable Inhabitant where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Const (UnifiedVariable variable)) where
+    synthetic = const (NonSimplifiable Nothing)
+
+instance Synthetic NonSimplifiable (Const StringLiteral) where
+    synthetic = const (NonSimplifiable . Just $ BuiltinHead)
+    {-# INLINE synthetic #-}
+
+instance Synthetic NonSimplifiable (Top sort) where
+    synthetic = const (NonSimplifiable Nothing)
+    {-# INLINE synthetic #-}
+
+data NonSimplifiableHead = ConstructorLikeHead
+                         | SortInjectionHead
+                         | BuiltinHead
```

Comment:
Why is `BuiltinHead` distinct from `ConstructorLikeHead`?

Reply:
Yeah, I guess this distinction didn't turn out to be necessary. I'll remove one of them.